### PR TITLE
add a missed sid check when adding login_hint

### PR DESCRIPTION
**Pull request by sameerag**
While adding login_hint in addHintParameters, need to check if sid is already populated. Adding the check.

### CODE REVIEW

File: lib/msal-core/src/UserAgentApplication.ts [hunk 1]
```diff
@@ -1133,7 +1133,7 @@ export class UserAgentApplication {
       // login_hint
       else {
         // login_hint is account.userName
-        if (!qParams[SSOTypes.LOGIN_HINT]  && account.userName && !Utils.isEmpty(account.userName)) {
+        if (!qParams[SSOTypes.SID] && !qParams[SSOTypes.LOGIN_HINT] && account.userName && !Utils.isEmpty(account.userName)) {
```
Comment (DarylThayil) [inline lib/msal-core/src/UserAgentApplication.ts]: this looks find, I wonder if factoring this into variable will add to readability

```javascript
const thingWeAreCheckingFor = predicate && predicate && predicate; 
if(!thingWeAreCheckingFor) {  ... }
```

Comment (sameerag) [reply lib/msal-core/src/UserAgentApplication.ts]: no ... so this is a test branch, hence I haven't yet sent it for a review. This is for the customer to test the change. Any style that simplifies readability, can always be incorporated.

File: lib/msal-core/src/UserAgentApplication.ts [hunk 2]
```diff
@@ -1122,18 +1122,18 @@ export class UserAgentApplication {
...
         }
       }
       // login_hint
       else {
         // login_hint is account.userName
-        if (!qParams[SSOTypes.LOGIN_HINT]  && account.userName && !Utils.isEmpty(account.userName)) {
+        if (!qParams[SSOTypes.LOGIN_HINT] && account.userName && !Utils.isEmpty(account.userName)) {
           qParams = Utils.addSSOParameter(SSOTypes.LOGIN_HINT, account.userName, qParams);
         }
       }
```
Comment (pkanher617) [inline lib/msal-core/src/UserAgentApplication.ts]: Should we be throwing an error here if they pass both? looks like we silently remove or don't add a parameter if both are present without saying anything to the caller.

File: lib/msal-core/src/UserAgentApplication.ts [hunk 3]
```diff
@@ -1122,18 +1122,18 @@ export class UserAgentApplication {
...
-        if (!qParams[SSOTypes.SID]  && !qParams[SSOTypes.LOGIN_HINT]) {
+        if (!qParams[SSOTypes.LOGIN_HINT]) {
           qParams = Utils.addSSOParameter(SSOTypes.SID, account.sid, qParams);
         }
       }
       // login_hint
       else {
         // login_hint is account.userName
-        if (!qParams[SSOTypes.LOGIN_HINT]  && account.userName && !Utils.isEmpty(account.userName)) {
+        if (!qParams[SSOTypes.LOGIN_HINT] && account.userName && !Utils.isEmpty(account.userName)) {
```
Comment (pkanher617) [inline lib/msal-core/src/UserAgentApplication.ts]: You can simplify this by moving the '!qParams[SSOTypes.LOGIN_HINT]' check to line 1126.

File: lib/msal-core/src/UserAgentApplication.ts [hunk 4]
```diff
@@ -1122,18 +1122,18 @@ export class UserAgentApplication {
     const account: Account = accountObj || this.getAccount();
 
     // This is a final check for all queryParams added so far; preference order: sid > login_hint
-    // sid cannot be passed along with login_hint, hence we check both are not populated yet in queryParameters so far
-    if (account) {
+    // sid cannot be passed along with login_hint or domain_hint, hence we check both are not populated yet in queryParameters so far
```
Comment (pkanher617) [inline lib/msal-core/src/UserAgentApplication.ts]: The comment says sid cannot be passed along with login_hint or domain_hint, but where do we actually check this and remove the hint if sid is available?

File: lib/msal-core/src/UserAgentApplication.ts [hunk 5]
```diff
@@ -1122,18 +1122,18 @@ export class UserAgentApplication {
...
+    if (account && !qParams[SSOTypes.SID]) {
       // sid
       if (account.sid && serverReqParams.promptValue === PromptState.NONE) {
-        if (!qParams[SSOTypes.SID]  && !qParams[SSOTypes.LOGIN_HINT]) {
+        if (!qParams[SSOTypes.LOGIN_HINT]) {
```
Comment (pkanher617) [inline lib/msal-core/src/UserAgentApplication.ts]: You can simplify this by moving the '!qParams[SSOTypes.LOGIN_HINT]' check to line 1126.

File: lib/msal-core/src/UserAgentApplication.ts (hunk 2 above)
Comment (sameerag) [reply lib/msal-core/src/UserAgentApplication.ts]: That check is done at a different place in populateQueryParams.

Comment (sameerag) [reply lib/msal-core/src/UserAgentApplication.ts]: This is only fall back when the user does not pass anything, that is why we check if it is already populated.

File: lib/msal-core/src/UserAgentApplication.ts (hunk 3 above)
Comment (sameerag) [reply lib/msal-core/src/UserAgentApplication.ts]: +1. I am also checking if we can optimize more of hints code to simplify all checks we make regarding hints.

File: lib/msal-core/src/UserAgentApplication.ts (hunk 5 above)
Comment (sameerag) [reply lib/msal-core/src/UserAgentApplication.ts]: +1

**Comment by DarylThayil** [comment]
Is this still a test branch? or is it good to be added to dev?

**Comment by sameerag** [comment]
@DarylThayil go ahead and finish the review. This is good to be added to dev now. I incorporated your review comments.

**Pull request merged**